File: lectures/lecture09/python/stack.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None


class StackLinkedList:
    """
    linked list implementation of stack
    - similar to singly linked list
    """

    def __init__(self):
        self.top = None

    def push(self, data):
        """
        add element to top of stack, O(1)
        """
        if not self.top:
            self.top = Node(data)
        else:
            node = Node(data)
            node.prev = self.top
            self.top = node
    
    def pop(self):
        """
        pop element from top of stack, O(1)
        """
        if not self.top:
            raise Exception("Cannot pop from empty stack")
        data = self.top.data
        self.top = self.top.prev
        return data

    def __str__(self):
        """
        string representation of stack
        """
        s = ""
        node = self.top
        while node is not None:
            s += str(node.data) + " "
            node = node.prev
        return s


class StackArrayList:
    """
    array list implementation of stack
    """

    def __init__(self):
        self.arr = []

    def push(self, data):
        self.arr.insert(0, data)

    def pop(self):
        return self.arr.pop(0)

    def __str__(self):
        return str(self.arr)


if __name__ == "__main__":

    stack = StackLinkedList()
    stack.push(5)
    stack.push(3)
    stack.push(7)
    print(str(stack))  # 7, 3, 5
    stack.pop()
    print(str(stack))  # 3, 5

    stack = StackArrayList()
    stack.push(5)
    stack.push(3)
    stack.push(7)
    print(str(stack))  # 7, 3, 5
    stack.pop()
    print(str(stack))  # 3, 5

File: courses/cmpt225/lecture09/cpp/src/stack.cpp
#include "stack.h"
#include "node.h"

#include <exception>
#include <string>

template <class Type>
StackLinkedList<Type>::StackLinkedList() : top(nullptr)
{
}

template <class Type>
StackLinkedList<Type>::~StackLinkedList()
{
    Node<Type> *node = top;
    while (node)
    {
        Node<Type> *next = node->next;
        delete node;
        node = next;
    }
}

template <class Type>
void StackLinkedList<Type>::push(Type data)
{
    if (!top)
    {
        top = new Node<Type>(data);
    }
    else
    {
        Node<Type> *node = new Node<Type>(data);
        node->prev = top;
        top = node;
    }   
}

template <class Type>
Type StackLinkedList<Type>::pop()
{
    if (!top)
    {
        // "Cannot pop from empty stack"
        throw std::exception();
    }

    Type data = top->data;
    Node<Type> *prev = top->prev;
    delete top;
    top = prev;

    return data;
}

template <>
std::string StackLinkedList<int>::to_string() const
{
    std::string s;

    Node<int> *node = top;

    while (node)
    {
        s += std::to_string(node->data) + " ";
        node = node->prev;
    }

    return s;
}

template <>
std::string StackLinkedList<std::string>::to_string() const
{
    std::string s;

    Node<std::string> * node = top;

    while (node)
    {
        s += node->data + " ";
        node = node->prev;
    }

    return s;
}

template <class Type>
void StackArrayList<Type>::push(Type data)
{
    arr.insert(arr.begin(), data);
}

template <class Type>
Type StackArrayList<Type>::pop()
{
    if (arr.empty())
    {
        // "Cannot pop from empty stack"
        throw std::exception();
    }

    Type data = arr[0];

    arr.erase(arr.begin());

    return data;
}

template <>
std::string StackArrayList<int>::to_string() const
{
    std::string s;

    for (auto it : arr)
    {
        s += std::to_string(it) + " ";
    }

    return s;
}

template <>
std::string StackArrayList<std::string>::to_string() const
{
    std::string s;

    for (auto it : arr)
    {
        s += it + " ";
    }

    return s;
}

template class StackLinkedList<int>;
template class StackLinkedList<std::string>;

template class StackArrayList<int>;
template class StackArrayList<std::string>;